File: sdks/python/nexus_ipc/decorators.py
from __future__ import annotations

import functools
from typing import Any, Callable, TypeVar

from .client import NexusClient

F = TypeVar("F", bound=Callable[..., Any])


def nexus_checkpoint(
    agent_id: str,
    channel: str,
    run_id: str = "decorated-run",
    client: NexusClient | None = None,
) -> Callable[[F], F]:
    sdk_client = client or NexusClient.connect()

    def decorator(func: F) -> F:
        @functools.wraps(func)
        def wrapper(*args: Any, **kwargs: Any) -> Any:
            result = func(*args, **kwargs)
            state = result if isinstance(result, dict) else {"result": result}
            sdk_client.checkpoint(
                agent_id=agent_id,
                run_id=run_id,
                channel=channel,
                state=state,
                summary=f"Checkpoint after {func.__name__}",
            )
            return result

        return wrapper  # type: ignore[return-value]

    return decorator


def nexus_tool(
    name: str,
    agent_id: str = "tool-agent",
    run_id: str = "tool-run",
    channel: str | None = None,
    client: NexusClient | None = None,
) -> Callable[[F], F]:
    sdk_client = client or NexusClient.connect()
    tool_channel = channel or f"tool:{name}"

    def decorator(func: F) -> F:
        @functools.wraps(func)
        def wrapper(*args: Any, **kwargs: Any) -> Any:
            sdk_client.publish_delta(
                channel=tool_channel,
                agent_id=agent_id,
                run_id=run_id,
                delta={"tool": name, "phase": "started"},
                durable=False,
            )
            try:
                result = func(*args, **kwargs)
            except Exception as exc:
                sdk_client.checkpoint(
                    agent_id=agent_id,
                    run_id=run_id,
                    channel=tool_channel,
                    state={"tool": name, "error": type(exc).__name__},
                    summary=f"Tool {name} failed",
                )
                raise
            sdk_client.checkpoint(
                agent_id=agent_id,
                run_id=run_id,
                channel=tool_channel,
                state={"tool": name, "result": result},
                summary=f"Tool {name} finished",
            )
            return result

        return wrapper  # type: ignore[return-value]

    return decorator

File: sdks/typescript/src/decorators.ts
import { NexusClient } from "./client";

export function checkpointStep<TState extends Record<string, unknown>, TResult>(
  client: NexusClient,
  options: { agentId: string; runId: string; channel: string; summary?: string },
  step: (state: TState) => Promise<TResult> | TResult,
): (state: TState) => Promise<TResult> {
  return async (state: TState) => {
    const result = await step(state);
    const output = typeof result === "object" && result !== null ? (result as Record<string, unknown>) : { result };
    await client.checkpoint({
      agentId: options.agentId,
      runId: options.runId,
      channel: options.channel,
      state: output,
      summary: options.summary ?? "Checkpointed wrapped step",
    });
    return result;
  };
}

export function wrapTool<TArgs extends unknown[], TResult>(
  client: NexusClient,
  options: { name: string; agentId: string; runId: string; channel?: string },
  tool: (...args: TArgs) => Promise<TResult> | TResult,
): (...args: TArgs) => Promise<TResult> {
  const channel = options.channel ?? `tool:${options.name}`;
  return async (...args: TArgs) => {
    await client.publishDelta({
      channel,
      agentId: options.agentId,
      runId: options.runId,
      delta: { tool: options.name, phase: "started" },
      durable: false,
    });
    try {
      const result = await tool(...args);
      await client.checkpoint({
        agentId: options.agentId,
        runId: options.runId,
        channel,
        state: { tool: options.name, result },
        summary: `Tool ${options.name} finished`,
      });
      return result;
    } catch (error) {
      await client.checkpoint({
        agentId: options.agentId,
        runId: options.runId,
        channel,
        state: { tool: options.name, error: error instanceof Error ? error.name : "Error" },
        summary: `Tool ${options.name} failed`,
      });
      throw error;
    }
  };
}

